### PR TITLE
first recipe ExitWaveReconstruction

### PR DESCRIPTION
**Pull request by din14970**
Checklist
- [x ] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml"
- [x ] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x ] Source is from official source
- [x ] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [x ] If static libraries are linked in, the license of the static library is packaged.
- [x ] Build number is 0
- [x ] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x ] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by jan-janssen** [comment]
Can you try adding `ocl-icd` https://anaconda.org/conda-forge/ocl-icd ?

**Comment by jan-janssen** [comment]
To fix the mac build:
```
    - khronos-opencl-icd-loader      # [win or osx]
    - ocl_icd_wrapper_apple      # [osx]
    - ocl-icd      # [linux]
```
https://github.com/conda-forge/pyopencl-feedstock/blob/master/recipe/meta.yaml#L43

**Comment by din14970** [comment]
Thanks Jan, I was already trying `ocl-icd` and will wait to see the result on the linux build, then I will add the correct changes for osx and windows as you suggest

**Comment by din14970** [comment]
hmm still the same problem. On my Arch linux laptop I was able to compile no problem, just had to install `opencl-header` from the package index, which was still separate from `libclc` and `ocl-icd`. Maybe a conda package needs to be made from <https://github.com/KhronosGroup/OpenCL-Headers>?

**Comment by jan-janssen** [comment]
Both ocl and pocl include the header files, but they are in slightly different locations:

- pocl `${PREFIX}/share/pocl/include`
- ocl `${PREFIX}/include/OpenCL`

So maybe you have to add something like `IFLAGS="-I${PREFIX}/include"`

**Comment by jan-janssen** [comment]
The option `USE_GNUPLOT` does not seem to exist: 
```
CMake Warning:
  Manually-specified variables were not used by the project:

    USE_GNUPLOT

```

**Comment by jan-janssen** [comment]
It is strange it is looking for: 
```
fatal error: CL/cl.hpp: No such file or directory
```
But this file is not included in 

https://github.com/OCL-dev/ocl-icd/tree/master/khronos-headers/CL

I only find it in

https://www.archlinux.org/packages/extra/any/opencl-headers/files/

Nt sure where this is coming from.

**Comment by jan-janssen** [comment]
I guess we need a package for https://github.com/KhronosGroup/OpenCL-CLHPP/tree/master

**Comment by jan-janssen** [comment]
Try adding `clhpp`

**Comment by din14970** [comment]
but looking there it seems on master cl.hpp is depricated. Maybe an older version can be specified somewhere?

**Comment by jan-janssen** [comment]
At least the compiling works now: https://github.com/conda-forge/staged-recipes/pull/12306 

It fails with:

```
mkdir: cannot create directory `/home/conda/staged-recipes/build_artifacts/exitwavereconstruction_1596631648501/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/bin': File exists
Traceback (most recent call last):
  File "/home/conda/.ci_support/build_all.py", line 133, in <module>
    build_all(args.recipes_dir, args.arch)
  File "/home/conda/.ci_support/build_all.py", line 66, in build_all
    build_folders(recipes_dir, new_comp_folders, arch, channel_urls)
  File "/home/conda/.ci_support/build_all.py", line 123, in build_folders
    conda_build.api.build([recipe], config=get_config(arch, channel_urls))
  File "/opt/conda/lib/python3.7/site-packages/conda_build/api.py", line 209, in build
    notest=notest, need_source_download=need_source_download, variants=variants)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 2864, in build_tree
    notest=notest,
  File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 1995, in build
    cwd=src_dir, stats=build_stats)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/utils.py", line 404, in check_call_env
    return _func_defaulting_env_to_os_environ('call', *popenargs, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/utils.py", line 384, in _func_defaulting_env_to_os_environ
    raise subprocess.CalledProcessError(proc.returncode, _args)
subprocess.CalledProcessError: Command '['/bin/bash', '-o', 'errexit', '/home/conda/staged-recipes/build_artifacts/exitwavereconstruction_1596631648501/work/conda_build.sh']' returned non-zero exit status 1.

```

maybe you need something like `mkdir -p` . I leave this to you.

**Comment by din14970** [comment]
@conda-forge/staged-recipes ready for review

**Comment by jan-janssen** [comment]
I guess you can reduce the host dependencies to: 
```
  build:
     - {{ compiler('cxx') }}
     - cmake
     - make
   host:
     - libblas
     - fftw
     - libtiff
     - bzip2
     - boost-cpp
     - clblas
     - ocl-icd
     - clhpp
```

**Comment by din14970** [comment]
how do you find out which ones are unnecessary?

**Comment by jan-janssen** [comment]
I just tried to remove ones that are not listed in the make script

File: recipes/ExitWaveReconstruction/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    - boost-cpp
+    - ocl-icd  # [linux]
+    - ocl_icd_wrapper_apple  # [osx]
+    - clhpp
+    - cimg
+    - xorg-libx11
+  run:
+    - _openmp_mutex  # [linux]
+    - clblas  # [linux]
+
```
Comment (jan-janssen) [inline recipes/ExitWaveReconstruction/meta.yaml:36]:
```suggestion
    - xorg-libx11
    - lib

```

**Comment by jan-janssen** [comment]
@conda-forge/staged-recipes ready for review

File: recipes/CImg/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  url: https://github.com/dtschump/CImg/archive/v.{{ version }}.tar.gz 
+  sha256: 3ad10cb8842e945af208a05b8ed36f4428317ef754c6e341e4ca6ac70b8a8540
+
+build:
+  noarch: ''
+  noarch_python: false
```
Comment (chrisburr) [inline recipes/CImg/meta.yaml:14]:
```suggestion
  noarch: generic
```

File: recipes/CImg/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  number: 0
+
+requirements:
+  build:
+  host:
+  run:
+
+test:
+  imports:
```
Comment (chrisburr) [inline recipes/CImg/meta.yaml:23]: There should always be a test even if it's a simple existence test.
```suggestion
  commands:
    - test -f something
```

File: recipes/CImg/build.bat [hunk 4]
```diff
@@ -0,0 +1,6 @@
+@echo off 
+
+mkdir %PREFIX%/include/
```
Comment (chrisburr) [inline recipes/CImg/build.bat:3]: Is this the right place for headers on Windows?

* If it is, this file can be removed as `noarch: generic` packages are only build on Linux
* If it isn't, the package can't be `noarch: generic`

File: recipes/ExitWaveReconstruction/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    - boost-cpp
+    - ocl-icd  # [linux]
+    - ocl_icd_wrapper_apple  # [osx]
+    - clhpp
+    - cimg
+    - xorg-libx11
+  run:
+    - _openmp_mutex  # [linux]
+    - clblas  # [linux]
+    - zlib
```
Comment (chrisburr) [inline recipes/ExitWaveReconstruction/meta.yaml:36]: I think this is needed in host instead

File: recipes/CImg/build.bat (hunk 4 above)
Comment (din14970) [reply recipes/CImg/build.bat:3]: I can only assume it is correct, I simply copied the build instructions from <https://anaconda.org/dtk-forge/cimg>. The guy who made this recipe also created the package. I will remove the `noarch` lines in meta.yaml

File: recipes/ExitWaveReconstruction/meta.yaml (hunk 5 above)
Comment (din14970) [reply recipes/ExitWaveReconstruction/meta.yaml]: removing clblas from run caused build to fail, the others I have moved to host

File: recipes/CImg/meta.yaml (hunk 3 above)
Comment (din14970) [reply recipes/CImg/meta.yaml:21]: test added

File: recipes/CImg/meta.yaml (hunk 2 above)
Comment (din14970) [reply recipes/CImg/meta.yaml]: see below

File: recipes/CImg/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+build:
+  number: 0
+
+requirements:
+  build:
+  host:
+  run:
+
+test:
+  imports:
```
Comment (jan-janssen) [inline recipes/CImg/meta.yaml:21]: You can remove this line

Comment (din14970) [reply recipes/CImg/meta.yaml:21]: You mean the imports line right? Should I just remove the `requirements` section altogether since it is empty?

Comment (jan-janssen) [reply recipes/CImg/meta.yaml:21]: I just mean `imports` in the test section.

File: recipes/ExitWaveReconstruction/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    - libtiff
+    - clblas
+    - boost-cpp
+    - ocl-icd  # [linux]
+    - ocl_icd_wrapper_apple  # [osx]
+    - clhpp
+    - cimg
+    - xorg-libx11
+    - _openmp_mutex  # [linux]
```
Comment (beckermr) [inline recipes/ExitWaveReconstruction/meta.yaml:34]: see the docs on openmp: https://conda-forge.org/docs/maintainer/knowledge_base.html#openmp

**Comment by beckermr** [comment]
you need to remove the openmp mutext package as well.

File: recipes/ExitWaveReconstruction/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
```
Comment (jan-janssen) [inline recipes/ExitWaveReconstruction/meta.yaml:36]:
```suggestion

```

**Pull request merged**